core/example/longrange/ewald.cpp

```diff
@@ -0,0 +1,270 @@
+              if (k - lx > r_max) continue;
+              for (auto ky = 0; ky <= k_max_int[1]; ++ky)
+              {
+                //check if cell within r_max distance in x+y
+                k = sqrt( (double)kx * (double)kx * lx * lx +
+                  (double)ky * (double)ky * ly * ly);
+                if (k - lx  > r_max) continue;
+                for (auto kz = 0; kz <= k_max_int[2]; ++kz)
+                {
+                  //Exclude self-energy term when kx=ky=kz
+                  if ( kx == 0 && ky == 0 && kz == 0 && i == idx) continue;
+                  //check if cell within r_max distance in x+y+z
+                  k = sqrt( (double)kx * (double)kx * lx * lx +
+                            (double)ky * (double)ky * ly * ly +
+                            (double)kz * (double)kz * lz * lz );
+                  if (k - lx > r_max) continue;
+                  //check if particle distance is less than r_max
+                  double scal = (d[0] + (double)kx * lx) * (d[0] + (double)kx * lx) +
+                    (d[1] + (double)ky * ly) * (d[1] + (double)ky * ly) +
+                    (d[2] + (double)kz * lz) * (d[2] + (double)kz * lz);
+                  scal = sqrt(scal);
+                  if (scal > r_max) 
+                    continue;
+                  //Compute real-space energy contribution of interaction
+                  Ur_part += qiqj * erfc(alpha * scal)/scal;
+                }
+              }
+            }
+          }
+        Ur_part *= 0.5;
+        p(idx) += Ur_part;
+      },
+      Ur);
+
+  // computation reciprocal-space contribution
+  int k_int = std::ceil(_k_max);
+  Kokkos::parallel_reduce( Kokkos::RangePolicy<ExecutionSpace>(0,8*k_int*k_int*k_int+12*k_int*k_int+6*k_int+1), KOKKOS_LAMBDA(int idx, double& Uk_part)
+      {
+      double kk;
+      double kr;
+      double coeff;
+      double sum_re = 0.0;
+      double sum_im = 0.0;
+      int kx = idx % (2 * k_int + 1) - k_int;
+      int ky = idx % (4 * k_int * k_int + 4 * k_int + 1) / (2 * k_int + 1) - k_int;
+      int kz = idx / (4 * k_int * k_int + 4 * k_int + 1) - k_int;
+
+      if (kx == 0 && ky == 0 && kz == 0) return;
+      kk = kx*kx + ky*ky + kz*kz;
+      if (kk > k_max*k_max) return;
+      coeff = 2.0 / (lx * lx) * exp( - PI_SQ / (alpha * alpha * lx * lx) * kk) / kk;
+      for (auto j = 0; j < n_max; ++j)
+      {
+      kr = 2.0 * PI * (kx*r(j,0) + ky*r(j,1) + kz*r(j,2)) / lx;
+      sum_re += q(j)*cos(kr);
+      sum_im += q(j)*sin(kr);
+      }
+      for (auto j = 0; j < n_max; ++j)
+      {
+        kr = 2.0 * PI * (kx*r(j,0) + ky*r(j,1) + kz*r(j,2)) / lx;
+        double re = sum_re * cos(kr);
+        double im = sum_im * sin(kr);
+
+        Kokkos::atomic_add(&p(j),q(j)*coeff*(re + im) * lx / (4.0*PI));
+
+        Uk_part += q(j)*coeff*(re + im) * lx/(4.0*PI);
+      }
+      }, Uk);
+
+  Kokkos::parallel_reduce( Kokkos::RangePolicy<ExecutionSpace>(0,n_max), KOKKOS_LAMBDA(int idx, double& Uk_part)
```

Comment:
what is this computing? total potential?

Reply:
Yes, this seems to compute the same quantity as in the previous loop. Removed.